crates/oci-distribution/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+        if challenge_opt.is_none() {
+            // This means that no challenge was present, even though the header was present.
+            // Since we do not handle basic auth, it could be the case that the upstream service
+            // is in compatibility mode with a Docker v1 registry.
+            return Ok(());
+        }
+
+        // Right now, we do read-only auth.
+        let pull_perms = format!("repository:{}:pull", image.repository());
+        let challenge = challenge_opt.unwrap()[0].clone();
+        let realm = challenge.realm.unwrap();
+        let service = challenge.service.unwrap();
+        let scope = pull_perms.to_owned();
+
+        // TODO: At some point in the future, we should support sending a secret to the
+        // server for auth. This particular workflow is for read-only public auth.
+        let auth_res = cli
+            .get(&realm)
+            .query(&[("service", service), ("scope", scope)])
+            .send()
+            .await?;
+
+        match auth_res.status() {
+            reqwest::StatusCode::OK => {
+                let docker_token: DockerToken = auth_res.json().await?;
+                self.token = Some(docker_token);
+                Ok(())
+            }
+            _ => {
+                let reason = auth_res.text().await?;
+                Err(failure::format_err!("failed to authenticate: {}", reason))
+            }
+        }
+    }
+
+    /// Pull a manifest from the remote OCI Distribution service.
+    ///
+    /// If the connection has already gone through authentication, this will
+    /// use the bearer token. Otherwise, this will attempt an anonymous pull.
+    pub async fn pull_manifest(&self, image: &Reference) -> OciResult<String> {
+        // We unwrap right now because this try_from literally cannot fail.
+        let cli = reqwest::Client::new();
+        let url = image.to_v2_manifest_url();
+        let request = cli.get(&url);
+
+        let mut headers = reqwest::header::HeaderMap::new();
+        headers.insert("Accept", "application/vnd.docker.distribution.manifest.v2+json,application/vnd.docker.distribution.manifest.list.v2+json".parse().unwrap());
+
+        if let Some(bearer) = self.token.as_ref() {
+            headers.insert("Authorization", bearer.bearer_token().parse().unwrap());
+        }
+
+        let res = request.headers(headers).send().await?;
+
+        let status = res.status();
+        if res.status().is_client_error() {
+            // According to the OCI spec, we should see an error in the message body.
+            let err = res.json::<OciEnvelope>().await?;
+            // FIXME: This should not have to wrap the error.
+            return Err(format_err!("{} on {}", err.errors[0], url));
+        } else if status.is_server_error() {
+            return Err(format_err!("Server error at {}", url));
+        }
+        let text = res.text().await?;
+        Ok(text)
+    }
+}
+
+#[derive(serde::Deserialize)]
+struct DockerToken {
```

Comment:
nit: Personally I would not use the `Docker*` nomenclature.

Reply:
OciToken? Who standardizes this token? I thought it was Docker in this case (not OCI)

Reply:
It is a registry token, so I would use `RegistryToken` for this.